src/rshim.c

```diff
@@ -2073,42 +2116,56 @@ int rshim_set_drop_mode(rshim_backend_t *bd, int value)
 {
   int old_value;
   int rt = 0;
+  int has_bd_lock;
+
+  has_bd_lock = !pthread_mutex_trylock(&bd->mutex);
 
-  pthread_mutex_lock(&bd->mutex);
   old_value = (int)bd->drop_mode;
   value = !!value;
   if (value == old_value) {
-    pthread_mutex_unlock(&bd->mutex);
-    return -EALREADY;
+    if (has_bd_lock)
+      pthread_mutex_unlock(&bd->mutex);
+    return 0;
   }
 
-  bd->drop_mode = 0;
-  if (bd->enable_device && bd->enable_device(bd, value ? false : true))
-    bd->drop_mode = 1;
-  else
-    bd->drop_mode = value;
+  if (has_bd_lock)
+    pthread_mutex_unlock(&bd->mutex);
 
-  if (bd->drop_mode)
-    bd->drop_pkt = 1;
-  else
-    rshim_fifo_sync(bd, true);
-  pthread_mutex_unlock(&bd->mutex);
   /*
    * Check if another endpoint driver has already attached to the
    * same rshim device before enabling it.
    */
-  if (!bd->drop_mode) {
+  if (!value) {
     rshim_lock();
-    pthread_mutex_lock(&bd->mutex);
+    has_bd_lock = !pthread_mutex_trylock(&bd->mutex);
     if (rshim_access_check(bd)) {
       RSHIM_WARN("rshim%d is not accessible\n", bd->index);
       bd->drop_mode = 1;
       rt = -EACCES;
+      return rt;
     }
-    pthread_mutex_unlock(&bd->mutex);
+    if (has_bd_lock)
+      pthread_mutex_unlock(&bd->mutex);
     rshim_unlock();
   }
-  return rt;
+
+  has_bd_lock = !pthread_mutex_trylock(&bd->mutex);
+
+  bd->drop_mode = 0;
+  if (bd->enable_device && bd->enable_device(bd, value ? false : true))
+    bd->drop_mode = 1;
+  else
+    bd->drop_mode = value;
+
+  if (bd->drop_mode)
+    bd->drop_pkt = 1;
+  else
+    rshim_fifo_sync(bd, true);
```

Comment:
Let's not to reorganize the code order in this commit unless there is good reason.

Reply:
The quick reason is when it's called in fuse thread to set drop mode to 0 when it's currently 1, it will first set bd->drop_mode to 1 and then do access check. But once bd->drop_mode is set to 1, another thread could start writing keep alive, causing the access check to fail. That's why switch the order helps. 

This is a tentative solution that aims to prove the fix. If we can come up with a less invasive fix that'll be better

Reply:
"But once bd->drop_mode is set to 1, another thread could start writing keep alive, causing the access check to fail. " 
Do you mean the peer rshim tries to write keep alive?  It's very rare unless we're trying to bring up the peer side at the same time. Even it happens, it's expected behavior because the peer side 'tries' to take over.

This change will cause problem when switching drop_mode from 1 to 0. The rshim resource has been released when drop_mode is 1. In such case, it'll try to read/write register without rshim resource mapped.

Reply:
>> Do you mean the peer rshim tries to write keep alive? 
No. I mean the main thread that writes KeepAlive magic to SP1 in the same rshim driver.

>> This change will cause problem when switching drop_mode from 1 to 0. The rshim resource has been released when drop_mode is 1. In such case, it'll try to read/write register without rshim resource mapped.
I got this part. I am trying to check why it seems doing fine in my test. And also trying to find a solution without this issue.

Reply:
The original code is fine. This block of code is done inside pthread_mutex_lock(), there is no other rshim write could be done (thus no way for the timer thread to send out keepalive).

Reply:
The block of code is not inside pthread_mutex_lock but inside pthread_mutex_trylock, which could fail if the thread running keep-alive already holds the bd->lock. This race condition explains why it succeeds sometimes but not other times.